manifests/init.pp

```diff
@@ -357,163 +357,184 @@
+  Array $docker_users                                 = [],
+  String $docker_group                                = $docker::params::docker_group,
+  Array $daemon_environment_files                     = [],
+  Variant[String,Hash,Undef] $repo_opt                = $docker::params::repo_opt,
+  Boolean $nowarn_kernel                              = $docker::params::nowarn_kernel,
+  Optional[String] $os                                = $docker::params::os,
+  Optional[String] $storage_devs                      = $docker::params::storage_devs,
+  Optional[String] $storage_vg                        = $docker::params::storage_vg,
+  Optional[String] $storage_root_size                 = $docker::params::storage_root_size,
+  Optional[String] $storage_data_size                 = $docker::params::storage_data_size,
+  Optional[String] $storage_min_data_size             = $docker::params::storage_min_data_size,
+  Optional[String] $storage_chunk_size                = $docker::params::storage_chunk_size,
+  Optional[String] $storage_growpart                  = $docker::params::storage_growpart,
+  Optional[String] $storage_auto_extend_pool          = $docker::params::storage_auto_extend_pool,
+  Optional[String] $storage_pool_autoextend_threshold = $docker::params::storage_pool_autoextend_threshold,
+  Optional[String] $storage_pool_autoextend_percent   = $docker::params::storage_pool_autoextend_percent,
+  Optional[String] $storage_config                    = $docker::params::storage_config,
+  Optional[String] $storage_config_template           = $docker::params::storage_config_template,
+  Optional[String] $storage_setup_file                = $docker::params::storage_setup_file,
+  Optional[String] $service_provider                  = $docker::params::service_provider,
+  Optional[String] $service_config                    = $docker::params::service_config,
+  Optional[String] $service_config_template           = $docker::params::service_config_template,
+  Optional[String] $service_overrides_template        = $docker::params::service_overrides_template,
+  Optional[Boolean] $service_hasstatus                = $docker::params::service_hasstatus,
+  Optional[Boolean] $service_hasrestart               = $docker::params::service_hasrestart,
 
 ) inherits docker::params {
 
-  validate_string($version)
-  validate_re($::osfamily, '^(Debian|RedHat|Archlinux|Gentoo)$',
-              'This module only works on Debian or Red Hat based systems or on Archlinux as on Gentoo.')
-  validate_bool($manage_kernel)
-  validate_bool($manage_package)
-  validate_bool($manage_service)
-  validate_array($docker_users)
-  validate_array($daemon_environment_files)
-  validate_array($log_opt)
-  validate_bool($tls_enable)
-  validate_bool($ip_forward)
-  validate_bool($iptables)
-  validate_bool($ip_masq)
-  if $icc != undef {
-    validate_bool($icc)
+  if $version {
+    assert_type(String[1], $version)
```

Comment:
Not 100% sure, but I think you can put this check in the type declaration itself, like - Optional[String[1]] $version - Will save you the extra couple of lines of the assert type here :)